<a href="https://colab.research.google.com/github/LatiefDataVisionary/machine-learning-basic-dicoding/blob/main/%5BKlasifikasi%5D_Submission_Akhir_BMLP_Lathif_Ramadhan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [35]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, f1_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV


# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [15]:
from google.colab import drive

In [16]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [22]:
df = pd.read_csv('/content/drive/MyDrive/Project Machine Learning Dicoding/hasil_clustering.csv')
df

ID  Year_Birth    Income   Kidhome  Teenhome Dt_Customer   Recency  \
0      5524   -0.985345  0.235327 -0.825218 -0.929894  2012-09-04  0.307039   
1      2174   -1.235733 -0.235826  1.032559  0.906934  2014-03-08 -0.383664   
2      4141   -0.317643  0.773633 -0.825218 -0.929894  2013-08-21 -0.798086   
3      6182    1.268149 -1.022732  1.032559 -0.929894  2014-02-10 -0.798086   
4      5324    1.017761  0.241519  1.032559 -0.929894  2014-01-19  1.550305   
...     ...         ...       ...       ...       ...         ...       ...   
2235  10870   -0.150717  0.358568 -0.825218  0.906934  2013-06-13 -0.107383   
2236   4001   -1.903435  0.470064  2.890335  0.906934  2014-06-10  0.237969   
2237   7270    1.017761  0.189106 -0.825218 -0.929894  2014-01-25  1.446700   
2238   8235   -1.068807  0.679035 -0.825218  0.906934  2014-01-24 -1.419719   
2239   9405   -1.235733  0.024838  1.032559  0.906934  2012-10-15 -0.314594   

      MntWines  MntFruits  MntMeatProducts  ...  Education_Master  \
0     0.983781   1.551577         1.679702  ...               0.0   
1    -0.870479  -0.636301        -0.713225  ...               0.0   
2     0.362723   0.570804        -0.177032  ...               0.0   
3    -0.870479  -0.560857        -0.651187  ...               0.0   
4    -0.389085   0.419916        -0.216914  ...               0.0   
...        ...        ...              ...  ...               ...   
2235  1.203678   0.419916         0.066692  ...               0.0   
2236  0.303291  -0.661449        -0.606873  ...               0.0   
2237  1.795020   0.545656         0.221789  ...               0.0   
2238  0.368666   0.092992         0.208495  ...               1.0   
2239 -0.653555  -0.586005        -0.469501  ...               0.0   

      Education_PhD  Marital_Status_Alone  Marital_Status_Divorced  \
0               0.0                   0.0                      0.0   
1               0.0                   0.0                      0.0   
2               0.0                   0.0                      0.0   
3               0.0                   0.0                      0.0   
4               1.0                   0.0                      0.0   
...             ...                   ...                      ...   
2235            0.0                   0.0                      0.0   
2236            1.0                   0.0                      0.0   
2237            0.0                   0.0                      1.0   
2238            0.0                   0.0                      0.0   
2239            1.0                   0.0                      0.0   

      Marital_Status_Married  Marital_Status_Single  Marital_Status_Together  \
0                        0.0                    1.0                      0.0   
1                        0.0                    1.0                      0.0   
2                        0.0                    0.0                      1.0   
3                        0.0                    0.0                      1.0   
4                        1.0                    0.0                      0.0   
...                      ...                    ...                      ...   
2235                     1.0                    0.0                      0.0   
2236                     0.0                    0.0                      1.0   
2237                     0.0                    0.0                      0.0   
2238                     0.0                    0.0                      1.0   
2239                     1.0                    0.0                      0.0   

      Marital_Status_Widow  Marital_Status_YOLO  cluster  
0                      0.0                  0.0        2  
1                      0.0                  0.0        1  
2                      0.0                  0.0        2  
3                      0.0                  0.0        2  
4                      0.0                  0.0        2  
...                    ...                  ...      ...  
2235            

In [28]:
df = df.drop('Dt_Customer', axis=1)

In [29]:
df.isnull().sum()

ID                         0
Year_Birth                 0
Income                     0
Kidhome                    0
Teenhome                   0
Recency                    0
MntWines                   0
MntFruits                  0
MntMeatProducts            0
MntFishProducts            0
MntSweetProducts           0
MntGoldProds               0
NumDealsPurchases          0
NumWebPurchases            0
NumCatalogPurchases        0
NumStorePurchases          0
NumWebVisitsMonth          0
AcceptedCmp3               0
AcceptedCmp4               0
AcceptedCmp5               0
AcceptedCmp1               0
AcceptedCmp2               0
Complain                   0
Z_CostContact              0
Z_Revenue                  0
Response                   0
Year                       0
Month                      0
Day                        0
Education_Basic            0
Education_Graduation       0
Education_Master           0
Education_PhD              0
Marital_Status_Alone       0
Marital_Status_Divorced    0
Marital_Status_Married     0
Marital_Status_Single      0
Marital_Status_Together    0
Marital_Status_Widow       0
Marital_Status_YOLO        0
cluster                    0
dtype: int64

# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [30]:
X = df.drop('cluster', axis=1)
y = df['cluster']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [31]:
print("Dimensi X_train:", X_train.shape)
print("Dimensi X_test:", X_test.shape)
print("Dimensi y_train:", y_train.shape)
print("Dimensi y_test:", y_test.shape)

Dimensi X_train: (1792, 40)
Dimensi X_test: (448, 40)
Dimensi y_train: (1792,)
Dimensi y_test: (448,)


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [32]:
# Inisialisasi model Random Forest
model = RandomForestClassifier(random_state=42)

# Latih model menggunakan data latih
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Tulis narasi atau penjelasan algoritma yang Anda gunakan.

Kode yang saya gunakan memanfaatkan algoritma **Random Forest Classifier** untuk melakukan klasifikasi. Sederhananya, bayangkan kita memiliki data pelanggan dan ingin mengelompokkannya ke dalam *cluster* berdasarkan karakteristik mereka. Random Forest akan membantu dalam tugas ini.

**Cara Kerja:**

1. **Kekuatan Bersama (Ensemble Learning)**: Random Forest termasuk dalam keluarga algoritma *Ensemble Learning*. Artinya, ia menggabungkan prediksi dari banyak model sederhana (**Decision Tree/Pohon Keputusan**) untuk menghasilkan prediksi yang lebih akurat dan andal. Ibaratnya, kita meminta pendapat dari banyak ahli, bukan hanya satu, untuk mendapatkan keputusan yang lebih baik.

2. **Membangun Pohon Keputusan:**

  * Random Forest menciptakan banyak pohon keputusan.
  * Setiap pohon dilatih dengan data yang berbeda, dipilih secara acak dari data asli (**bootstrap sampling**). Seperti membagi data menjadi beberapa bagian dan memberikannya ke pohon-pohon berbeda untuk dipelajari.
  * Saat membangun setiap pohon, hanya beberapa fitur yang dipilih secara acak untuk dipertimbangkan di setiap percabangan (**feature randomness**). Ini membuat pohon-pohon memiliki karakteristik yang berbeda dan mengurangi kemungkinan salah prediksi karena terlalu fokus pada data latihan.

**3. Membuat Prediksi:**

  * Ketika ada data baru yang ingin diprediksi, data tersebut akan diproses oleh setiap pohon keputusan di dalam *forest*.
  * Setiap pohon memberikan prediksinya sendiri.
  * Untuk tugas klasifikasi (seperti dalam kode ini), prediksi akhir ditentukan melalui **voting**. Kelas/cluster yang mendapatkan suara terbanyak dari pohon-pohon tersebut akan menjadi prediksi akhirnya.


## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

In [33]:
# Lakukan prediksi pada data uji
y_pred = model.predict(X_test)

In [36]:
# Hitung metrik evaluasi
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='weighted') # Gunakan average='weighted' untuk multi-class

In [37]:
# Cetak hasil evaluasi
print("Accuracy:", accuracy)
print("F1-Score:", f1)

Accuracy: 1.0
F1-Score: 1.0


**Arti Accuracy dan F1-Score 1.0**

* **Accuracy 1.0**: Artinya model saya berhasil memprediksi semua data di data uji dengan benar. Tidak ada data yang salah diklasifikasikan.
* **F1-Score 1.0**: F1-Score adalah metrik yang menggabungkan precision dan recall.   Nilai 1.0 menunjukkan bahwa model saya memiliki precision dan recall yang sempurna, yang berarti:
  * **Precision 1.0**: Semua data yang diprediksi sebagai positif (milik kelas tertentu) memang benar-benar positif. Tidak ada False Positive.
  * **Recall 1.0**: Semua data yang sebenarnya positif berhasil diidentifikasi oleh model. Tidak ada False Negative.

Hasil ini mengindikasikan bahwa model saya memiliki performa yang sangat baik pada data uji. Model mampu memprediksi semua kelas dengan sempurna.

In [38]:
# Buat dan tampilkan confusion matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)

Confusion Matrix:
 [[125   0   0]
 [  0 156   0]
 [  0   0 167]]


**Penjelasan Elemen Matriks**

* **[125, 0, 0]**:
  * 125 data yang sebenarnya milik kelas 1 diprediksi dengan benar sebagai kelas 1.
  * 0 data yang sebenarnya milik kelas 1 diprediksi salah sebagai kelas 2.
  * 0 data yang sebenarnya milik kelas 1 diprediksi salah sebagai kelas 3.
* **[0, 156, 0]**:
  * 0 data yang sebenarnya milik kelas 2 diprediksi salah sebagai kelas 1.
  * 156 data yang sebenarnya milik kelas 2 diprediksi dengan benar sebagai kelas 2.
  * 0 data yang sebenarnya milik kelas 2 diprediksi salah sebagai kelas 3.
* [0, 0, 167]:
  * -0 data yang sebenarnya milik kelas 3 diprediksi salah sebagai kelas 1.
  * 0 data yang sebenarnya milik kelas 3 diprediksi salah sebagai kelas 2.
  * 167 data yang sebenarnya milik kelas 3 diprediksi dengan benar sebagai kelas 3.

**Kesimpulan**

Dari Confusion Matrix ini, dapat disimpulkan bahwa model saya memiliki performa yang **sangat baik** dalam mengklasifikasikan data.

* **Tidak ada False Positive atau False Negative**: Semua data diprediksi dengan benar ke dalam kelas yang seharusnya.
* **Diagonal Utama**: Nilai-nilai pada diagonal utama (125, 156, 167) menunjukkan jumlah prediksi yang benar untuk setiap kelas. Nilai-nilai ini tinggi, yang menunjukkan performa model yang baik.
* **Nilai di Luar Diagonal**: Nilai-nilai di luar diagonal utama adalah 0, yang berarti tidak ada kesalahan klasifikasi.

Hasil ini konsisten dengan Accuracy dan F1-Score 1.0 yang saya dapatkan sebelumnya. Model saya mampu membedakan dan mengklasifikasikan ketiga kelas dengan sempurna pada data uji.

In [41]:
# Tampilkan classification report untuk detail lebih lanjut
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       125
           1       1.00      1.00      1.00       156
           2       1.00      1.00      1.00       167

    accuracy                           1.00       448
   macro avg       1.00      1.00      1.00       448
weighted avg       1.00      1.00      1.00       448



**Istilah dalam Classification Report**

* **Precision**: Proporsi data yang diprediksi positif dan benar-benar positif.
  * Rumus: `Precision = True Positive / (True Positive + False Positive)`
* **Recall**: Proporsi data positif yang berhasil diidentifikasi oleh model.
  * Rumus: `Recall = True Positive / (True Positive + False Negative)`
* F1-Score: Rata-rata harmonik dari precision dan recall. Memberikan keseimbangan antara keduanya.
  * Rumus: `F1-Score = 2 * (Precision * Recall) / (Precision + Recall)`
* **Support**: Jumlah data aktual untuk setiap kelas.

**Interpretasi Classification Report**

Report ini menunjukkan performa model untuk setiap kelas (`0`, `1`, dan `2`) dan secara keseluruhan.

* **Kelas 0**:

  * Precision: 1.00 (Semua data yang diprediksi sebagai kelas 0 memang benar-benar kelas 0)
  * Recall: 1.00 (Semua data yang sebenarnya kelas 0 berhasil diidentifikasi)
  * F1-Score: 1.00 (Performa sempurna untuk kelas 0)
  * Support: 125 (Jumlah data aktual untuk kelas 0)
* **Kelas 1 dan 2**: Pola yang sama dengan kelas 0, menunjukkan performa sempurna untuk kedua kelas ini juga.
* **Accuracy**: 1.00 (Akurasi keseluruhan model, yang berarti semua data diprediksi dengan benar)

* **macro avg**: Rata-rata precision, recall, dan F1-Score untuk semua kelas tanpa mempertimbangkan jumlah data per kelas.

* **weighted avg**: Rata-rata precision, recall, dan F1-Score untuk semua kelas dengan mempertimbangkan jumlah data per kelas.

**Kesimpulan**

Classification Report ini menegaskan bahwa model saya memiliki **performa yang sangat baik** dalam mengklasifikasikan data ke dalam tiga kelas. Semua metrik (`precision`, `recall`, `F1-Score`) bernilai **1.00** untuk semua kelas, menunjukkan tidak ada kesalahan klasifikasi pada data uji.

Tulis hasil evaluasi algoritma yang digunakan, jika Anda menggunakan 2 algoritma, maka bandingkan hasilnya.

## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [ ]:
#Type your code here

## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [ ]:
#Type your code here

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.